### PR TITLE
Parse incrementally (closes #117)

### LINKED ISSUE #117: Make SSE parsing incremental
Comment by @viktorklang: "Since you pull when there is no data, then you could make the parsing incremental rather than reparsing everything until there is data. I.e. have a mutable builder which you then apply lines to, as they arrive, and then have it create SSEs when there is something to send."

### PR DESCRIPTION
**Pull request by hseeberger**

### CODE REVIEW

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala [hunk 1]
```diff
@@ -22,6 +22,58 @@ import scala.util.control.NonFatal
...
+      _size += 5 + value.length
+    }
+
+    def setEventType(value: String): Unit = {
+      eventType = value
+      _size += 6 + value.length
+    }
+
+    def setId(value: String): Unit = {
+      id = value
```
Comment (viktorklang) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: Note that if you are overwriting a previous value you "need" to decrement the _size accordingly (or rather decrement/increment size based on size of new vs old)

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala [hunk 2]
```diff
@@ -22,6 +22,58 @@ import scala.util.control.NonFatal
 
 private object ServerSentEventParser {
 
+  private class Builder {
```
Comment (viktorklang) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: If you make the builder `final` you give the JVM optimizer more info to inline the calls to the methods.

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala (hunk 1 above)
Comment (hseeberger) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: Not sure, depends on the meaning of `size`.

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala (hunk 2 above)
Comment (hseeberger) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: +1

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala [hunk 3]
```diff
@@ -70,31 +122,34 @@ private final class ServerSentEventParser(maxEventSize: Int) extends GraphStage[
...
-        }
-        lines = newLines
-        linesSize = newSize
+
+      private val builder = new Builder()
+
+      override def onPush() = grab(in) match {
+        case "" => // A server-sent event is terminated with a new line, i.e. an empty line
+          if (builder.hasData) push(out, builder.build()) else pull(in)
+          builder.reset()
```
Comment (viktorklang) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: Should it be reset even if `pull(in)`?

**Comment by viktorklang** [comment]
:+1:

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala (hunk 3 above)
Comment (hseeberger) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: Good catch! I need to add suitable test data to catch this.

Comment (hseeberger) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: No, wait, this is correct. We are talking about an event without data: that must be ignored.

Comment (hseeberger) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: I'll add some comments.

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala (hunk 1 above)
Comment (hseeberger) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: For the parser it's probably more important to look at the memory size, not at the wire size. Hence you are right.

File: akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala [hunk 4]
```diff
@@ -22,6 +22,61 @@ import scala.util.control.NonFatal
 
 private object ServerSentEventParser {
 
+  private final class Builder {
+
+    private var data = Vector.empty[String]
```
Comment (viktorklang) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: I recommend switching to a mutable data structure for this now, as it is a mutable builder anyway.

Comment (hseeberger) [inline akka-sse/src/main/scala/de/heikoseeberger/akkasse/ServerSentEventParser.scala]: ACK

**Pull request merged**